### PR TITLE
Feature RM#64181: TAX / TAXTYPE / SEVERAL OBJECTS : Change the M2O to…

### PR DESCRIPTION
**Pull request by bsu-axelor**
… Tax/TaxLine into a M2M

### CODE REVIEW

File: axelor-account/src/main/java/com/axelor/apps/account/service/fixedasset/FixedAssetLineMoveServiceImpl.java [hunk 1]
```diff
@@ -756,37 +765,53 @@ public void generateSaleMove(
...
+      for (TaxLine taxLine : saleTaxLineSet) {
+
+        List<AccountManagement> creditAccountTwoList =
+            taxLine.getTax().getAccountManagementList().stream()
+                .filter(
+                    accountManagement ->
+                        accountManagement
+                            .getCompany()
+                            .getName()
+                            .equals(fixedAsset.getCompany().getName()))
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/java/com/axelor/apps/account/service/fixedasset/FixedAssetLineMoveServiceImpl.java:779]: accountManagement can have company other than fixedAsset.company.

File: axelor-account/src/main/resources/i18n/messages_fr.csv [hunk 2]
```diff
@@ -1353,7 +1353,7 @@
 "Default expense account","Compte de charge par défaut",,
 "Default income account","Compte de produit par défaut",,
 "Default service type","Type de service (DAS2) par défaut",,
-"Default tax","Taxe par défaut",,
+"Default taxes",,,
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/resources/i18n/messages_fr.csv:1356]: Please provide translation

File: axelor-account/src/main/resources/i18n/messages_fr.csv [hunk 3]
```diff
@@ -2992,7 +2995,7 @@
 "Some invoice lines analytic distributions are not correct, please correct them.","Certaines distributions analytiques sur les lignes de factures ne sont pas corrects. Veuillez les corriger.",,
 "Some move amounts have been changed since the imputation. Please remake the imputation.","Certains montants d’écritures ont été modifiés depuis l’imputation. Veuillez procéder à une nouvelle imputation.",,
 "Some selected MoveLines already have a proposal ReconcileGroup","Certaines lignes sélectionnées ont déjà une proposition de lettrage",,
-"Source Tax","Taxe à la source",,
+"Source Taxes",,,
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/resources/i18n/messages_fr.csv:2998]: Please provide translation

File: axelor-account/src/main/resources/i18n/messages_fr.csv [hunk 4]
```diff
@@ -3113,6 +3116,7 @@
 "Tax required on move line","Taxe obligatoire sur les lignes d'écriture",,
 "Tax standard rate for clearance","T.V.A. standard pour apurements",,
 "Tax standard rate for irrecoverables","T.V.A. standard pour irrécouvrables",,
+"Tax type",,,
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/resources/i18n/messages_fr.csv:3119]: Please provide translation

File: axelor-account/src/main/resources/views/MoveTemplateLine.xml [hunk 5]
```diff
@@ -135,7 +135,7 @@
 
   <action-record name="action-move-template-line-record-account-onchange"
     model="com.axelor.apps.account.db.MoveTemplateLine">
-    <field name="tax" expr="eval: account?.defaultTax"/>
+    <field name="tax" expr="eval: account?.defaultTaxSet[0]"/>
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/resources/views/MoveTemplateLine.xml:138]: It is setting tax to null when set is empty

File: axelor-base/src/main/resources/domains/TaxType.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  xsi:schemaLocation="http://axelor.com/xml/ns/domain-models http://axelor.com/xml/ns/domain-models/domain-models_6.1.xsd">
+
+  <module name="account" package="com.axelor.apps.account.db"/>
+
+  <entity name="TaxType" cacheable="true">
+
+    <string name="name" title="Name" required="true"/>
+    <string name="code" title="Code" required="true" unique="true"/>
+    <integer name="typeSelect" title="Type" selection="tax.type.type.select" required="true"
+      nullable="true"/>
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-base/src/main/resources/domains/TaxType.xml:13]: Without nullable, required will not be effective as typeSelect will set to 0 when no value is selected.

File: axelor-account/src/main/java/com/axelor/apps/account/service/moveline/massentry/MoveLineMassEntryRecordServiceImpl.java [hunk 7]
```diff
@@ -160,7 +160,7 @@ public void resetPartner(MoveLineMassEntry moveLine, MoveLineMassEntry newMoveLi
       moveLine.setPartnerFullName(null);
       moveLine.setMovePartnerBankDetails(null);
       moveLine.setVatSystemSelect(null);
-      moveLine.setTaxLine(null);
+      moveLine.setTaxLineSet(null);
```
Comment (aol-axelor) [inline axelor-account/src/main/java/com/axelor/apps/account/service/moveline/massentry/MoveLineMassEntryRecordServiceImpl.java:163]: Please set an empty set instead. Please also change if `null` is set elsewhere

File: axelor-account/src/main/java/com/axelor/apps/account/web/FixedAssetController.java [hunk 8]
```diff
@@ -108,19 +111,22 @@ public void disposal(ActionRequest request, ActionResponse response) throws Axel
...
-                    ((Integer) ((HashMap<String, Object>) context.get("saleTaxLine")).get("id"))
-                        .longValue());
+      if (context.get("saleTaxLineSet") != null) {
+        Collection<Map<String, Object>> saleTaxLineMapSet =
+            (Collection<Map<String, Object>>) context.get("saleTaxLineSet");
+        TaxLineRepository taxLineRepository = Beans.get(TaxLineRepository.class);
+        for (Map<String, Object> saleTax : saleTaxLineMapSet) {
+          TaxLine saleTaxLine =
+              taxLineRepository.find(Long.parseLong(saleTax.get("id").toString()));
+          saleTaxLineSet.add(saleTaxLine);
```
Comment (aol-axelor) [inline axelor-account/src/main/java/com/axelor/apps/account/web/FixedAssetController.java:128]: Please use a stream with `forEach` at the end here

File: axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java [hunk 9]
```diff
@@ -375,24 +378,58 @@ public void updateTaxEquiv(ActionRequest request, ActionResponse response) {
...
 
-        if (fiscalPosition != null && taxLine != null) {
-          taxEquiv =
-              Beans.get(FiscalPositionService.class).getTaxEquiv(fiscalPosition, taxLine.getTax());
+        if (fiscalPosition != null && CollectionUtils.isNotEmpty(taxLineSet)) {
+          Map<TaxLine, TaxLine> taxMap = new HashMap<>();
+          if (CollectionUtils.isNotEmpty(moveLine.getTaxLineBeforeReverseSet())) {
+            for (TaxLine taxLineBeforeRev : moveLine.getTaxLineBeforeReverseSet()) {
+              taxEquiv =
+                  Beans.get(FiscalPositionService.class)
```
Comment (aol-axelor) [inline axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java:397]: Please inject service outside of loop

File: axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java [hunk 10]
```diff
@@ -375,24 +378,58 @@ public void updateTaxEquiv(ActionRequest request, ActionResponse response) {
...
-            taxLine =
-                Beans.get(TaxService.class).getTaxLine(taxEquiv.getToTax(), moveLine.getDate());
+          for (TaxLine taxLine : taxLineSet) {
+            if (taxMap.containsKey(taxLine)) {
+              taxLineBeforeReverseSet.add(taxMap.get(taxLine));
+            } else {
+              taxLineBeforeReverseSet.add(taxLine);
+            }
+            taxEquiv =
+                Beans.get(FiscalPositionService.class)
```
Comment (aol-axelor) [inline axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java:420]: Please inject service outside of loop

File: axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java [hunk 11]
```diff
@@ -375,24 +378,58 @@ public void updateTaxEquiv(ActionRequest request, ActionResponse response) {
...
+                    .getTaxEquiv(fiscalPosition, taxLine.getTax());
+            if (taxEquiv != null) {
+              TaxLine newTaxLine =
+                  Beans.get(TaxService.class).getTaxLine(taxEquiv.getToTax(), moveLine.getDate());
```
Comment (aol-axelor) [inline axelor-account/src/main/java/com/axelor/apps/account/web/MoveLineController.java:424]: Please inject service outside of loop

File: axelor-account/src/main/resources/i18n/messages_fr.csv (hunk 2 above)
[the remark this answers is not in the document]
Comment (aol-axelor) [reply axelor-account/src/main/resources/i18n/messages_fr.csv:1375]: `Taxes par défaut`

File: axelor-account/src/main/resources/i18n/messages_fr.csv [hunk 12]
```diff
@@ -1,5 +1,5 @@
 "key","message","comment","context"
-"${label} move lines","Lignes d'écritures : ${label}",,
+"${label} move lines",,,
```
Comment (aol-axelor) [inline axelor-account/src/main/resources/i18n/messages_fr.csv:2]: Why are so many translations being deleted in this file ? Please restore them

File: axelor-account/src/main/resources/i18n/messages_fr.csv (hunk 3 above)
[the remark this answers is not in the document]
Comment (aol-axelor) [reply axelor-account/src/main/resources/i18n/messages_fr.csv:3029]: `Taxes à la source`

File: axelor-base/src/main/java/com/axelor/apps/base/service/tax/AccountManagementServiceImpl.java [hunk 13]
```diff
@@ -151,14 +162,17 @@ public TaxEquiv getProductTaxEquiv(
...
+  protected Set<Tax> getProductTax(
       Product product, Company company, FiscalPosition fiscalPosition, boolean isPurchase)
       throws AxelorException {
-    Tax generalTax = this.getProductTax(product, company, isPurchase);
-    Tax tax = fiscalPositionService.getTax(fiscalPosition, generalTax);
+    Set<Tax> generalTaxSet = this.getProductTax(product, company, isPurchase);
+    Set<Tax> taxSet = Sets.newHashSet();
+    for (Tax generalTax : generalTaxSet) {
+      taxSet.add(fiscalPositionService.getTax(fiscalPosition, generalTax));
+    }
```
Comment (aol-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/tax/AccountManagementServiceImpl.java:172]: Please use a stream here

File: axelor-base/src/main/java/com/axelor/apps/base/service/tax/FiscalPositionServiceImpl.java [hunk 14]
```diff
@@ -47,4 +50,16 @@ public TaxEquiv getTaxEquiv(FiscalPosition fiscalPosition, Tax tax) {
...
+  @Override
+  public Set<TaxEquiv> getTaxEquivSet(FiscalPosition fiscalPosition, Set<Tax> taxSet) {
+    Set<TaxEquiv> taxEquivSet = new HashSet<>();
+    for (Tax tax : taxSet) {
+      TaxEquiv taxEquiv = getTaxEquiv(fiscalPosition, tax);
+      if (ObjectUtils.notEmpty(taxEquiv)) {
+        taxEquivSet.add(taxEquiv);
+      }
+    }
+    return taxEquivSet;
```
Comment (aol-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/tax/FiscalPositionServiceImpl.java:63]: Please return the result of a stream

File: axelor-base/src/main/java/com/axelor/apps/base/service/tax/TaxService.java [hunk 15]
```diff
@@ -85,25 +91,96 @@ public TaxLine getTaxLine(Tax tax, LocalDate localDate) throws AxelorException {
...
+  public Set<TaxLine> getTaxLineSet(Set<Tax> taxSet, LocalDate localDate) throws AxelorException {
+
+    if (CollectionUtils.isEmpty(taxSet)) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_CONFIGURATION_ERROR, I18n.get(BaseExceptionMessage.TAX_2));
+    }
+    Set<TaxLine> taxLineSet = Sets.newHashSet();
+
+    for (Tax tax : taxSet) {
+      if (ObjectUtils.notEmpty(tax.getActiveTaxLine())) {
```
Comment (aol-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/tax/TaxService.java:110]:
```suggestion
      if (tax.getActiveTaxLine() != null) {
```

File: axelor-sale/src/main/resources/i18n/messages_fr.csv [hunk 16]
```diff
@@ -622,7 +630,7 @@
 "Total price","Prix total",,
 "Total quantity sold","Quantité totale vendue",,
 "Total stock","Stock total",,
-"Total tax","Total T.V.A.",,
+"Total tax","Total Tax",,
```
Comment (aol-axelor) [inline axelor-sale/src/main/resources/i18n/messages_fr.csv:633]:
```suggestion
"Total tax","Total taxe",,

```

File: axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineTaxServiceImpl.java [hunk 17]
```diff
@@ -134,43 +146,46 @@ protected TaxPaymentMoveLine generateTaxPaymentMoveLine(
...
+    Set<TaxLine> taxLineSet = invoiceMoveLine.getTaxLineSet();
+    List<TaxPaymentMoveLine> taxPaymentMoveLineList = new ArrayList<>();
+    for (TaxLine taxLine : taxLineSet) {
+      BigDecimal vatRate = taxLine.getValue();
+
+      BigDecimal baseAmount = BigDecimal.ZERO;
+      if (BigDecimal.ZERO.compareTo(vatRate) != 0) {
+        baseAmount =
+            (invoiceMoveLine.getCredit().add(invoiceMoveLine.getDebit()))
+                .divide(vatRate.divide(new BigDecimal(100)), 2, BigDecimal.ROUND_HALF_UP);
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineTaxServiceImpl.java:158]: Can't set scale from currency as scale=2 is specified in domain and if currency has numberOfDecimals > 2, then validation error will be triggered. so i am utilizing global constant.

File: axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineFinancialDiscountServiceImpl.java [hunk 18]
```diff
@@ -473,16 +502,28 @@ public Map<Tax, Pair<BigDecimal, BigDecimal>> getFinancialDiscountTaxMap(MoveLin
...
     for (InvoiceLineTax invoiceLineTax : invoice.getInvoiceLineTaxList()) {
+      TaxLine taxLine = invoiceLineTax.getTaxLine();
+      List<InvoiceLine> invoiceLineList = invoiceLineTax.getInvoice().getInvoiceLineList();
+      long noOfLines =
+          invoiceLineList.stream().filter(il -> il.getTaxLineSet().contains(taxLine)).count();
+      long count =
+          invoiceLineList.stream()
+              .filter(il -> il.getTaxLineSet().contains(taxLine))
+              .flatMap(il -> il.getTaxLineSet().stream())
+              .count();
```
[the remark this answers is not in the document]
Comment (bsu-axelor) [reply axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineFinancialDiscountServiceImpl.java:519]: Once a operation is called on a stream, the stream is considered consumed or closed so we can't reuse it.

File: changelogs/unreleased/feature-m2o-tax-line-to-m2m-in-several-objects.yml [hunk 19]
```diff
@@ -0,0 +1,3 @@
+---
+title : "TAX / TAXTYPE / SEVERAL OBJECTS : Change the M2O to Tax/TaxLine into a M2M"
```
Comment (aol-axelor) [inline changelogs/unreleased/feature-m2o-tax-line-to-m2m-in-several-objects.yml:2]:
```suggestion
title : "TAX : Add a multi tax feature"
```

**Pull request merged**